Granite Vision Support

alex-jw-brooks opened the pull request:
# What does this PR do?

This PR adds compatibility for IBM's upcoming Granite Vision models (which are based on LLava Next). The main changes here are:
- The vision feature layer, which is currently expected to be an integer can now also be a list of integers; if a list of integers are provided, the image features are the concatenated before applying the feature selection strategy
- The validation which breaks visual encoders with no CLS (discussed a bit [here](https://github.com/huggingface/transformers/pull/35262)) is removed. I did add a warning if the feature packing explodes with the `default` strategy as well.

This change was applied in a lot of places to make the checks for repository consistency happy + the config consistent, but the multimodal granite models are instances of `LlavaNextForConditionalGeneration`. I added a test for each changed model to ensure that things don't blow up if a list of vision feature layers is provided, but if there is another path forward that is preferred to changing several models at the same time to add compatibility with llava next, I'm happy to revise this PR as needed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts, @qubvel, @zucchini-nlp

alex-jw-brooks committed:
Add multimodal granite support

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

Support multiple image feature layres

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Remove failing validation for visual encoders with no cls

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Update llava based models / configs to support list of feature layers

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add tests for multiple feature layers

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

zucchini-nlp reviewed (commented):
Great, thanks for propagating the changes to all llava models. I think we can also modify vipllava for the sake of consistency

Also, left one question for cases when we have a list `vision_feature_select_strategy` with `default` strategy. WDYT, since that seems to be the most intuitive behavior? Or is the Multimodal Granite not cropping `CLS` tokens?

alex-jw-brooks committed:
Use conditional instead of except for misaligned feature shapes

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

alex-jw-brooks commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -275,25 +281,38 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
+        if isinstance(vision_feature_layer, int):
+            selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
+        else:
+            hs_pool = [image_outputs.hidden_states[layer_idx] for layer_idx in vision_feature_layer]
+            selected_image_feature = torch.cat(hs_pool, dim=-1)
```
That's a great point! The IBM granite multimodal models use `full` as the vision feature selection strategy with siglip (no CLS) as the vision encoder, so I hadn't thought about what the behavior of `default` should be, but I completely agree with you.

I took a look at vipllava's feature selection ([this](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vipllava/modeling_vipllava.py#L294)) based on your other comment, and it seems like this is also the behavior of the CLS cropping there - I'll make the change!

alex-jw-brooks committed:
crop cls from each hidden state

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

alex-jw-brooks commented on src/transformers/models/video_llava/modeling_video_llava.py:
```diff
@@ -364,46 +370,64 @@ def _merge_input_ids_with_visual_features(
+            raise ValueError(f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}")
 
         image_outputs = self.image_tower(pixel_values_images, output_hidden_states=True)
-        image_outputs = image_outputs.hidden_states[vision_feature_layer].squeeze(1)
```
@zucchini-nlp, do you know why there is a squeeze here? It looks like it was added intentionally, but it seems like if it actually happens, it blows up the validation later on.

More concretely, in the tiny model for the unit tests, the unsqueezed features are of shape `[5, 2, 32]`, so if we slice out 1 per layer for CLS, this squeeze does happen, bringing it down to `[5, 32]`. This changes the validation [here](https://github.com/huggingface/transformers/blob/53f9b96b07d2ccca35d32d7250ca9392d3c0f967/src/transformers/models/video_llava/modeling_video_llava.py#L587) to check against the `# of tiles * hidden size` instead of `# of tiles * num hidden layers` (after applying the strategy).

For now, I've removed the squeeze, both here and from the video features just to be consistent, but if this like the wrong thing to do, please let me know 🙂

zucchini-nlp replied:
Hmm indeed the squeezing seems unnecessary. I believe it was a result of copying the method from another llava model, and since `squeeze()` didn't had any extra dimension to remove all was good until today

I am oke with removing it

alex-jw-brooks commented:
Thanks a lot for the quick review @zucchini-nlp 😄 For `vipllava`, do you mean allowing the `vision_feature_layers` to be of type `Union[int, List[int]]`, allowing a strategy, or something else?

zucchini-nlp commented:
@alex-jw-brooks 

> allowing the vision_feature_layers to be of type Union[int, List[int]]

exactly, this is what I meant so we can be consistent for all llava models :)

alex-jw-brooks committed:
Fix formatting

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Support single vision feature int in vipllava

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks commented:
Awesome, thank you for the clarification @zucchini-nlp 😄  I made the change in vipllava and rebased this PR, it should be ready for another look when you have a moment!

zucchini-nlp commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -275,31 +281,46 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
         Returns:
             image_features (`torch.Tensor`): Image feature tensor of shape `(num_images, image_length, embed_dim)`).
         """
+        if vision_feature_select_strategy not in ["default", "hidden"]:
```
A typo? Can be only "full" and "default", but we can add "hidden" if needed. Though not sure what that means

alex-jw-brooks replied:
It sure is, thank you for catching that! 😓 Fixed :)

zucchini-nlp reviewed (approved):
Thanks for iterating, LGTM! Left a tiny comment, seems to be typo :)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35579). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

alex-jw-brooks committed:
Fix typo in vision feature selection strategy validation

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

zucchini-nlp commented:
Cool, thanks a lot! Let's get one more review from the core maintainer and then we'll merge

ArthurZucker reviewed (commented):
LGTM, but we need to add a `granite_multimodal.md` with model details!

ArthurZucker commented:
We would also want to add an integration test with expected values!

alex-jw-brooks committed:
Add tentative integration test for granite vision models

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

ArthurZucker commented:
yep completely possible!

alex-jw-brooks committed:
Add granite vision docs

Replace multimodal granite refs with granite vision

Add granite vision / llava next alias

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

alex-jw-brooks commented:
Great, thanks so much @ArthurZucker! It looks like the models are actually going to be released as `granite vision` - I went ahead and updated references from `granite multimodal` -> `granite vision` and added a short description of the model + example to `docs/source/en/model_doc/granitevision.md`. It should be ready for another look when you have a moment!

It seems like there will be a publication,  but it's not available quite yet - I'll be sure to open follow-up PRs for the integration test and doc updates (e.g., paper citation and model ID fix if the name ends up incorrect in the example) in the coming weeks as things are published 😄

ArthurZucker commented on docs/source/en/model_doc/granitevision.md:
```diff
@@ -0,0 +1,85 @@
+
+# prepare image and text prompt, using the appropriate prompt template
+url = "https://github.com/haotian-liu/LLaVA/blob/1a91fc274d7c35a9b50b3cb29c4247ae5837ce39/images/llava_v1_5_radar.jpg?raw=true"
+image = Image.open(requests.get(url, stream=True).raw)
```
```suggestion
```
you no longer need this! image is processed directly!

zucchini-nlp replied:
Btw, for that you also need to change the chat template as below:

```python
conversation = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": url},
            {"type": "text", "text": "What is shown in this image?"},
        ],
    },
]
inputs = processor.apply_chat_template(
    conversation,
    add_generation_prompt=True,
    tokenize=True,
    return_dict=True,
    return_tensors="pt"
).to("cuda")
```

alex-jw-brooks replied:
Perfect, thanks for all your help @ArthurZucker @zucchini-nlp - I've updated the example to use the url directly instead! 😄

ArthurZucker reviewed (approved):
1 nit and let's merge! 🤗

alex-jw-brooks committed:
Use image url in granitevision example

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

ArthurZucker commented:
Congrats on the merge! 🚀
